torch/nn/utils/stateless.py

```diff
@@ -37,21 +36,52 @@ def _setattr(self, name: str, value: Any) -> None:
-    if old_val not in replaced_tensors_map:
-        replaced_tensors_map[old_val] = new_val
-    elif replaced_tensors_map[old_val] is not new_val:
-        warnings.warn("functional_call was passed multiple values for tied weights. "
-                      "This behavior is deprecated and will be an error in future versions")
+def _create_tied_weights_map(module, params_and_buffers):
+    # creates a weight map of {tied_name: name_given_by_user} for all weights where one of their tied weights is passed
+    #
+    # The basic algorithm looks like:
+    #   - index all weights by their original tensor value to find tied weights
+    #     - when we encounter a weight not used by the user, we save it in a set (second element in the tuple)
+    #     - when we run into a weight used by the user, we save that separate from the set as the first element in the tuple
+    #     - ending map looks like {tensor: (name_given_by_user, set(all_tied_names)}
+    #   - then loop through the values of this map (name_given_by_user and set(all_tied_names))
+    #     - for each element of all_tied_names, add {tied_name: name_given_by_user} to a new map
+
+    names = params_and_buffers.keys()
+    weight_to_name_and_tied_names: Dict[torch.Tensor, Tuple[Optional[str], Set[str]]] = {}
+
+    def add_to_name_map(name, t):
+        if t in weight_to_name_and_tied_names:
+            first_seen_name = weight_to_name_and_tied_names[t][0]
+            if name in names and first_seen_name and params_and_buffers[name] is not params_and_buffers[first_seen_name]:
+                raise ValueError(f"functional_call got values for both {name} and {first_seen_name}, which are tied.")
+            elif name in names:
+                weight_to_name_and_tied_names[t] = (name, weight_to_name_and_tied_names[t][1])
+            else:
+                weight_to_name_and_tied_names[t][1].add(name)
+        else:
+            weight_to_name_and_tied_names[t] = (name, set()) if name in names else (None, {name})
+
+    for name, t in module.named_parameters(remove_duplicate=False):
+        add_to_name_map(name, t)
+
+    for name, t in module.named_buffers(remove_duplicate=False):
+        add_to_name_map(name, t)
 
+    # make {tied_name: name_given_by_user} from pairs of (name_given_by_user, set(all_tied_names))
+    tied_weights_to_given_name = {}
+    for name, tied_names in weight_to_name_and_tied_names.values():
+        if name is None:  # no mapping was passed for this tensor, use original tensor
+            continue
+        for tied_name in tied_names:
+            tied_weights_to_given_name[tied_name] = name
+    return tied_weights_to_given_name
 
-def _create_swap_params(params_and_buffers, replaced_tensors_map):
-    def _swap_parameters(module, tensor_name: str, full_path: str, tensor: Tensor) -> None:
+
+def _create_swap_params(params_and_buffers):
+    def _swap_parameters(module, tensor_name: str, full_path: str, tensor: Optional[Tensor]) -> None:
```

Comment:
What is tensor being used for?

Reply:
It's unused but previously it was the original weight. Since it was harder to get for tied weights, I passed None and had to update the signature to match that. I can add a patch on top of this to remove that parameter since it looks unused

Reply:
SGTM